### PR TITLE
feat: GDS/GDSII import for layer stacks, port sources, and port detectors

### PR DESCRIPTION
**Pull request by teevee112**
## Summary                                                      

Adds a GDS import pipeline so photonic chip layouts (GDSII files) can be                                                                                        
loaded directly into fdtdx simulations without manual geometry setup.
                                                                                                                                                                
### New public API (`fdtdx` namespace)                          
                                                                                                                                                                
- `GDSLayerSpec` — maps a GDS layer/datatype to a material name, thickness,                                                                                     
  z-base offset, and optional boolean etch list
- `GDSPortSpec` — maps a GDS port marker layer to a propagation axis and name                                                                                   
  prefix for auto-generated sources/detectors                                                                                                                   
- `GDSLayerObject` — `StaticMultiMaterialObject` subclass that holds extruded                                                                                   
  polygon geometry; voxel mask is rasterized via `polygons_to_mask`                                                                                             
- `gds_layer_stack(gds_source, cell_name, layers, materials, simulation_volume, gds_center)`                                                                    
  — builds one `GDSLayerObject` + 2 constraints (z position, xy size) per layer spec                                                                            
- `gds_layer_stack_from_component(component, ...)` — same, but accepts a                                                                                        
  gdsfactory `Component` (gdsfactory is an optional dependency)                                                                                                 
- `sources_from_gds_ports(...)` — creates `ModePlaneSource` objects from port                                                                                   
  marker polygons; 4 constraints per port (propagation position, transverse                                                                                     
  center, height size, height center)                                                                                                                           
- `detectors_from_gds_ports(...)` — same but produces `ModeOverlapDetector`                                                                                     
  objects                                                                                                                                                       
                                                                                                                                                                
### Design notes                                                                                                                                                
                                                                
- GDS encodes x/y geometry only; z-extrusion is the only supported axis.                                                                                        
  Attempting `axis != 2` on `GDSLayerObject.get_voxel_mask_for_shape` raises
  `ValueError` with an explanation.                                                                                                                             
- `gds_center` is required (no default) — callers must explicitly state which                                                                                   
  GDS coordinate maps to the simulation x/y centre to avoid silent misalignment.                                                                                
- Boolean etch (`etch_by`) subtracts one layer's polygons from another via                                                                                      
  `gdstk.boolean(..., "not")` before voxelization.                                                                                                              
- Port propagation axis is limited to 0 (x) or 1 (y); `propagation_axis=2`                                                                                      
  raises `ValueError`.                                                                                                                                          
- `polygons_to_mask` changed from stacking all masks in memory to an iterative                                                                                  
  OR reduction (constant memory for large polygon counts).                                                                                                      
- Internal helpers `_PortEntry` + `_iter_port_placements` are shared between                                                                                    
  the source and detector functions to avoid duplication.                                                                                                       
                                                                                                                                                                
### Tests                                                                                                                                                       
                                                                                                                                                                
- Unit tests covering axis properties, voxel mask shape/content, material                                                                                       
  mapping, `gds_layer_stack` (names, constraints, missing cell), boolean etch
  edge cases, gdsfactory integration path, sources and detectors (constraint                                                                                    
  count, error cases, name prefix, empty layer)                                                                                                                 
- Integration tests covering single-layer placement, two-layer z-stacking, and                                                                                  
  missing-cell error

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Create simulation material stacks directly from GDS layouts and export them via the public API.
  * Generate mode sources and overlap detectors from GDS port markers.
  * Expanded public GDS integration APIs and a utility to rasterize polygon geometry into 2D masks for placement.

* **Tests**
  * Extensive unit and integration tests covering GDS stacking, voxelization, port-derived sources/detectors, etch behavior, naming, and error cases.

[![Review Change Stack](https://storage.googleapis.com/coderabbit_public_assets/review-stack-in-coderabbit-ui.svg)](https://app.coderabbit.ai/change-stack/ymahlau/fdtdx/pull/323)
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

### CODE REVIEW

**Comment by ymahlau** [comment]
@coderabbitai review

File: src/fdtdx/objects/static_material/gds_layer_stack.py [hunk 1]
```diff
@@ -0,0 +1,587 @@
...
+from fdtdx.core.jax.pytrees import autoinit, frozen_field
+from fdtdx.core.wavelength import WaveCharacter
+from fdtdx.materials import Material, compute_ordered_names
+from fdtdx.objects.detectors.mode import ModeOverlapDetector
+from fdtdx.objects.object import RealCoordinateConstraint
+from fdtdx.objects.sources.mode import ModePlaneSource
+from fdtdx.objects.static_material.static import SimulationVolume, StaticMultiMaterialObject
+
+
+@dataclass
```
Comment (ymahlau) [inline src/fdtdx/objects/static_material/gds_layer_stack.py]: please mark dataclasses as frozen for best jax compatibility

File: src/fdtdx/objects/static_material/gds_layer_stack.py [hunk 2]
```diff
@@ -0,0 +1,587 @@
...
+
+    Each instance represents one GDS layer (layer/datatype pair) extruded uniformly
+    along ``axis``. The cross-sectional shape is described by ``polygons``, given in
+    GDS coordinate space (metres). The mapping from GDS space to the local grid is
+    controlled by ``gds_center``, which gives the GDS coordinate that coincides with
+    the x/y center of the simulation object.
+    """
+
+    #: List of (N, 2) vertex arrays for each polygon, in GDS metres.
+    polygons: list[np.ndarray] = frozen_field()
```
Comment (ymahlau) [inline src/fdtdx/objects/static_material/gds_layer_stack.py]: Sequence might be a better annotation here

File: src/fdtdx/objects/static_material/gds_layer_stack.py [hunk 3]
```diff
@@ -0,0 +1,587 @@
...
+    thickness: float = frozen_field()
+
+    @property
+    def horizontal_axis(self) -> int:
+        """Cross-section axis that is not the extrusion axis and not the vertical axis.
+
+        Returns:
+            int: 1 (y) when extrusion axis is 0 (x), otherwise 0 (x).
+        """
+        return 1 if self.axis == 0 else 0
```
Comment (ymahlau) [inline src/fdtdx/objects/static_material/gds_layer_stack.py:105]: Not part of this PR, but added a issue to refactor this behavior into seperate functions #325

File: src/fdtdx/objects/static_material/gds_layer_stack.py [hunk 4]
```diff
@@ -0,0 +1,587 @@
...
+            partial_grid_shape=entry.pgshape,
+            partial_real_shape=entry.prshape,
+        )
+        prop_constraint = RealCoordinateConstraint(
+            object=entry.name,
+            axes=(entry.prop_axis,),
+            sides=("-",),
+            coordinates=(entry.sim_prop_pos,),
+        )
+        trans_center_constraint = src.place_at_center(simulation_volume, axes=(entry.trans_axis,))
```
Comment (ymahlau) [inline src/fdtdx/objects/static_material/gds_layer_stack.py:524]: This assumes that the full simulation only has a single layer, correct? While that is usually valid, the docstring should mention this

File: src/fdtdx/objects/static_material/gds_layer_stack.py [hunk 5]
```diff
@@ -0,0 +1,587 @@
...
+        height_constraint = src.same_size(simulation_volume, axes=(height_axis,))
+        vert_center_constraint = src.same_position(simulation_volume, axes=(height_axis,))
```
Comment (ymahlau) [inline src/fdtdx/objects/static_material/gds_layer_stack.py]: I am actually not sure what happens when placing the source at the same position as the volume. Does this place it at the center? Maybe place_at_center would be easier and can be combined with constraint above into a single constraint

**Comment by ymahlau** [comment]
@teevee112 Apart from the minor comments above, this already looks good! As an additional comment, maybe you could add a small (!) gds file into tests/data and run some unit / integration tests with that.

File: src/fdtdx/objects/static_material/gds_layer_stack.py (hunk 1 above)
Comment (teevee112) [reply src/fdtdx/objects/static_material/gds_layer_stack.py]: done

File: src/fdtdx/objects/static_material/gds_layer_stack.py (hunk 2 above)
Comment (teevee112) [reply src/fdtdx/objects/static_material/gds_layer_stack.py]: done

File: src/fdtdx/objects/static_material/gds_layer_stack.py (hunk 4 above)
Comment (teevee112) [reply src/fdtdx/objects/static_material/gds_layer_stack.py:524]: done

File: src/fdtdx/objects/static_material/gds_layer_stack.py (hunk 5 above)
Comment (teevee112) [reply src/fdtdx/objects/static_material/gds_layer_stack.py]: done

**Comment by ymahlau** [comment]
This looks good to me, please resolve the conflict and then we can merge

**Comment by ymahlau** [comment]
perfect, thanks!
